src/xrc/src/api.rs

```diff
@@ -33,88 +33,109 @@ const STABLECOIN_BASES: &[&str] = &[DAI, USDC];
 const MIN_NUM_RATES_FOR_PRIVILEGED_CANISTERS: usize =
     if cfg!(feature = "ipv4-support") { 3 } else { 2 };
 
+#[derive(CandidType, Deserialize, Clone, Debug)]
+struct QueriedExchangeRateWithFailedExchanges {
+    queried_exchange_rate: QueriedExchangeRate,
+    failed_exchanges: Vec<Exchange>,
+}
+
 #[async_trait]
 trait CallExchanges {
     async fn get_cryptocurrency_usdt_rate(
         &self,
+        exchanges: &[&Exchange],
         asset: &Asset,
         timestamp: u64,
-    ) -> Result<QueriedExchangeRate, CallExchangeError>;
+    ) -> Result<QueriedExchangeRateWithFailedExchanges, CallExchangeError>;
 
     async fn get_stablecoin_rates(
         &self,
+        exchanges: &[&Exchange],
         symbols: &[&str],
         timestamp: u64,
-    ) -> Vec<Result<QueriedExchangeRate, CallExchangeError>>;
+    ) -> Vec<Result<QueriedExchangeRateWithFailedExchanges, CallExchangeError>>;
 }
 
 struct CallExchangesImpl;
 
 #[async_trait]
 impl CallExchanges for CallExchangesImpl {
     async fn get_cryptocurrency_usdt_rate(
-        &self,
-        asset: &Asset,
-        timestamp: u64,
-    ) -> Result<QueriedExchangeRate, CallExchangeError> {
-        let futures = EXCHANGES.iter().filter_map(|exchange| {
-            if !exchange.is_available() {
-                return None;
-            }
-
-            Some(call_exchange(
-                exchange,
-                CallExchangeArgs {
-                    timestamp,
-                    quote_asset: usdt_asset(),
-                    base_asset: asset.clone(),
-                },
-            ))
-        });
-        let results = join_all(futures).await;
-
-        let mut rates = vec![];
-        let mut errors = vec![];
-        for result in results {
-            match result {
-                Ok(rate) => rates.push(rate),
-                Err(err) => {
-                    ic_cdk::println!(
-                        "{} Timestamp: {}, Asset: {:?}, Error: {}",
-                        LOG_PREFIX,
-                        timestamp,
-                        asset,
-                        err,
-                    );
-                    errors.push(err);
-                }
-            }
-        }
-
-        if rates.is_empty() {
-            return Err(CallExchangeError::NoRatesFound);
-        }
-
-        Ok(QueriedExchangeRate::new(
-            asset.clone(),
-            usdt_asset(),
-            timestamp,
-            &rates,
-            rates.len() + errors.len(),
-            rates.len(),
-            None,
-        ))
+      &self,
+      exchanges: &[&Exchange],
+      asset: &Asset,
+      timestamp: u64,
+    ) -> Result<QueriedExchangeRateWithFailedExchanges, CallExchangeError> {
+      let futures = exchanges.iter().filter_map(|exchange| {
+          if !exchange.is_available() {
```

Comment:
Move the is_available check to each `handle_` function. The `filter_map` can simply be `map` then.